src/vmm/src/acpi.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+use acpi_tables::{rsdp::RSDP, sdt::GenericAddress, sdt::SDT};
+use vm_memory::{Address, Bytes, ByteValued, GuestAddress, GuestMemoryMmap};
+
+#[repr(packed)]
+struct LocalAPIC {
+    pub r#type: u8,
+    pub length: u8,
+    pub processor_id: u8,
+    pub apic_id: u8,
+    pub flags: u32,
+}
+
+#[repr(packed)]
+#[derive(Default)]
+struct IOAPIC {
+    pub r#type: u8,
+    pub length: u8,
+    pub ioapic_id: u8,
+    _reserved: u8,
+    pub apic_address: u32,
+    pub gsi_base: u32,
+}
+
+#[repr(packed)]
+#[derive(Default)]
+struct InterruptSourceOverride {
+    pub r#type: u8,
+    pub length: u8,
+    pub bus: u8,
+    pub source: u8,
+    pub gsi: u32,
+    pub flags: u16,
+}
+
+const MADT_CPU_ENABLE_FLAG: usize = 0;
+
+fn create_dsdt_table(_cpu_count: u8) -> SDT {
+    // DSDT
+    let dsdt = SDT::new(*b"DSDT", 36, 6, *b"FIRECR", *b"FCDSDT  ", 1);
+
+    // TODO: add CPU devices
+
+    dsdt
+}
+
+fn create_madt(_cpu_count: u8) -> SDT {
```

Comment:
I would create a separate structure for `Madt` and define this method on the struct if possible. Or at least I would create helper methods on that structure in order to avoid working directly with `SDT::append` and other low-level functions that also seem quite usafe.

The same comment for the other tables as well.